### ISSUE TITLE
twisted.conch.endpoints.SSHCommandClientEndpoint misbehaves if initialized without a password

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#6553]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)| @exarkun reported|
|-|-|
|Trac ID|trac#6553|
|Type|defect|
|Created|2013-06-05 15:49:25Z|
|Branch|https://github.com/twisted/twisted/tree/ssh-endpoint-no-password-6553|

The `password` parameter to `newConnection` is documented as accepting `None`, and `None` is even the default value.

However, if `None` is given and the server offers password authentication, the endpoint will blow up when it tries to use `None` as a password:

```
        Traceback (most recent call last):
          File "twisted/internet/defer.py", line 138, in maybeDeferred
            result = f(*args, **kw)
          File "twisted/conch/ssh/userauth.py", line 459, in tryAuth
            return f()
          File "twisted/conch/ssh/userauth.py", line 706, in auth_password
            d.addCallbacks(self._cbPassword, self._ebAuth)
          File "twisted/internet/defer.py", line 294, in addCallbacks
            self._runCallbacks()
        --- <exception caught here> ---
          File "twisted/internet/defer.py", line 576, in _runCallbacks
            current.result = callback(current.result, *args, **kw)
          File "twisted/conch/ssh/userauth.py", line 732, in _cbPassword
            self.askForAuth('password', '\x00' + NS(password))
          File "twisted/conch/ssh/common.py", line 27, in NS
            return struct.pack('!L',len(t)) + t
        exceptions.TypeError: object of type 'NoneType' has no len()
```

Instead, if the password is `None` the endpoint shouldn't even try password authentication.

<details><summary>Searchable metadata</summary>

```
trac-id__6553 6553
type__defect defect
reporter__exarkun exarkun
priority__normal normal
milestone__ 
branch__branches_ssh_endpoint_no_password_6553 branches/ssh-endpoint-no-password-6553
branch_author__therve therve
status__closed closed
resolution__fixed fixed
component__conch conch
keywords__ 
time__1370447365000000 1370447365000000
changetime__1373474477000000 1373474477000000
version__None None
owner__therve therve
cc__z3p
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)|@therve set owner to @therve|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_3"></a>|@therve commented|
|-|-|

(In [39096]) Branching to 'ssh-endpoint-no-password-6553'

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)|@therve removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|<img alt="Alex's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_5"></a>|Alex commented|
|-|-|

A small (probably incomplete), review:

* twisted/conch/endpoints.py
  * ``getPassword`` is documented as returning a ``Deferred``, but it can now return ``None``.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_6"></a>|@therve commented|
|-|-|

(In [39099]) Mention the new return type

Refs [#6553](https://github.com/twisted/twisted/issues/6553)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun set owner to @exarkun<br>@exarkun set status to `assigned`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_8"></a>|@exarkun set owner to @therve<br>@exarkun set status to `new`|
|-|-|

Thanks!  This looks pretty good, and I'd be happy to see it merged as-is (though I am still waiting for buildbot - you forgot to force a build!).  However, I do have one comment about `test_skipPasswordAuthentication`.  It seems like it should be possible to exercise the password authentication case without using any keys, just talking to a server that only offers password authentication.  If so, the test might get a little bit simpler.  If you can make that change, great.  If not (and buildbot is green), go ahead and merge anyway.  Thanks again.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_9"></a>|@therve set status to `closed`|
|-|-|

(In [39105]) Merge ssh-endpoint-no-password-6553 

Author: therve
Reviewer: exarkun
Fixes: [#6553](https://github.com/twisted/twisted/issues/6553)

Skip password authentication in SSHCommandClientEndpoint if the password is not
specified.
